app/models/concerns/host_common.rb

```diff
@@ -92,15 +92,28 @@ def puppetca_exists?
     !!(puppet_ca_proxy && puppet_ca_proxy.url.present?)
   end
 
-  # no need to store anything in the db if the entry is plain "puppet"
-  # If the system is using smart proxies and the user has run the smartproxy:migrate task
-  # then the puppetmaster functions handle smart proxy objects
+  def puppet_server_uri
+    return unless puppet_proxy
+    url = puppet_proxy.setting('Puppet', 'puppet_url')
```

Comment:
This line needs tests. We need at least one test for both cases (setting is present + not present).